File: frontend/src/backend/exchangeRateService.ts
import api from "./api";

export interface ExchangeRateDto {
  from: string;
  to: string;
  rate: number;
  available: boolean;
}

export const exchangeRateService = {
  getRate(from: string, to: string): Promise<ExchangeRateDto> {
    return api.get("/exchange-rate", { params: { from, to } }).then((r) => r.data);
  },

  getLatestRates(base: string): Promise<Record<string, number>> {
    return api.get("/exchange-rate/latest", { params: { base } }).then((r) => r.data);
  },
};

File: backend/src/main/java/Balio/web/model/services/ExchangeRateService.java
package Balio.web.model.services;

import java.math.BigDecimal;
import java.util.Map;

public interface ExchangeRateService {

    /**
     * Returns the exchange rate to convert 1 unit of {@code from} into {@code to}.
     * Returns null if the rate cannot be fetched.
     */
    BigDecimal getRate(String from, String to);

    /**
     * Returns latest rates for all common currencies relative to {@code base}.
     * Returns empty map on failure.
     */
    Map<String, BigDecimal> getLatestRates(String base);
}
